sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,401 @@
+    };
+};
+
+/** Build an ETag based on the given file's VFS stat. If the stat fails,
+ * returns the error and leaves stattag in any state; otherwise there's an etag
+ * in the stattag's field */
+static int stat_etag(char *filename, union stattag *stattag);
+
+/* These are almost but but not quite coap_handler_t -- they require the
+ * request to be already fully read and digested into the last argument. */
+
+static ssize_t coapfileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+static ssize_t coapfileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+/** Create a CoAP response for a given errno (eg. EACCESS -> 4.03 Forbidden
+ * etc., defaulting to 5.03 Internal Server Error) */
+static size_t coapfileserver_errno_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, int err);
+/** Return true if path/name is a directory.
+ *
+ * Currently assembling the values in a buffer, this could be done without a
+ * buffer if an openat / vfs_openat was present.
+ * */
+static bool entry_is_dir(char *path, char *name);
+
+ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx) {
+    struct coapfileserver_entry *entry = (struct coapfileserver_entry *)ctx;
+    struct requestdata request;
+    request.etag_sent = false;
+    request.blocknum2 = 0;
+    request.szx2 = CONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX;
+    /** Index in request.namebuf. Must not point at the last entry as that will be
+     * zeroed to get a 0-terminated string. */
+    size_t namelength = 0;
+    bool trailing_slash = false;
+    /** If no path component comes along at all, it'll count as a trailing
+     * slash no matter the trailing_slash value */
+    bool any_component = false;
+    uint8_t errorcode = COAP_CODE_INTERNAL_SERVER_ERROR;
+
+    uint8_t strip_remaining = entry->strip_path;
+    namelength += request.namebuf - strncpy(request.namebuf, entry->nameprefix, sizeof(request.namebuf) - 1);
+
+    coap_optpos_t opt;
+    bool is_first = true;
+    uint8_t *value;
+    ssize_t optlen;
+    while ((optlen = coap_opt_get_next(pdu, &opt, &value, is_first)) !=
+            -ENOENT)
+    {
+        if (optlen < 0) {
+            errorcode = COAP_CODE_BAD_REQUEST;
+            goto error;
+        }
+
+        is_first = false;
+        switch (opt.opt_num) {
+            case COAP_OPT_URI_PATH:
+                if (strip_remaining != 0) {
+                    strip_remaining -= 1;
+                    continue;
+                }
+                if (trailing_slash) {
+                    errorcode = COAP_CODE_BAD_REQUEST;
+                    goto error;
+                }
+                any_component = true;
+                if (optlen == 0) {
+                    trailing_slash = true;
+                    continue;
+                }
+                if (memchr(value, '0', optlen) != NULL ||
```

Comment:
Why filter out paths containing '0' like `nvm0/`?

Reply:
This should have been a '\0' -- CoAP paths can easily contain both \0 and slashes, but letting them lose on our file system API would have ill-effects and is thus stopped early.

Reply:
> This should have been a '\0' -- CoAP paths can easily contain both \0 and slashes, but letting them lose on our file system API would have ill-effects and is thus stopped early.

I don't think I understood..., stumbled into the same issue as @benpicco

Reply:
The CoAP request may contain characters that would have ill-effects on RIOT:
* A \0 would be ignored because RIOT uses C-style strings in its VFS -- a `GET /dir/file%00name` would go to `/dir/file` on the file system. (I don't see immediate security concerns through that, only confusion, but still, reason enough to reject the request).
* In the CoAP request there might be a slash -- as in `GET /bands/AC%2FDC`. This should serve a file named "AC/DC" in the directory "bands", because RIOT (like almost all other OSs) has no way of expressing paths whose components contain the path delimiter.

What I should have written was:

```suggestion
                if (memchr(value, '\0', optlen) != NULL ||
```

Reply:
Thanks! I understand now!